### PR TITLE
[Tests] Fix tests 1st ed.

### PR DESCRIPTION
**Pull request by zhzhcookie**
### PR Category
[OP Test]

### Type of Change
[Tests]

### Description
1. Modify test cases, reduce similar type test cases, increase test cases of different dimensions and size ranges, and try to unify the test shapes of similar operators.
2. Fix --device=cpu.

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### Performance
image

### CODE REVIEW

File: tests/accuracy_utils.py [hunk 1]
```diff
@@ -26,37 +26,83 @@
 sizes_pow_2 = [2**d for d in range(4, 11, 2)]
 sizes_noalign = [d + 17 for d in sizes_pow_2]
 sizes_1d = sizes_one + sizes_pow_2 + sizes_noalign
-sizes_2d_nc = [1, 16, 64, 1000]
-sizes_2d_nr = [1, 5, 1024]
+sizes_2d_nc = [1] if TO_CPU else [1, 16, 64, 1000]
+sizes_2d_nr = [1] if TO_CPU else [1, 5, 1024]
```
Comment (iclementine) [inline tests/accuracy_utils.py:30]: `TO_CPU` indicates to compute the reference value on CPU, so use less and smaller sizes to reduce the runtime?

File: tests/accuracy_utils.py [hunk 2]
```diff
@@ -26,37 +26,83 @@
...
 
 UT_SHAPES_1D = list((n,) for n in sizes_1d)
 UT_SHAPES_2D = list(itertools.product(sizes_2d_nr, sizes_2d_nc))
-POINTWISE_SHAPES = [(1024, 1024), (16, 1024, 256), (16, 128, 64, 64), (20, 320, 15)]
+POINTWISE_SHAPES = (
+    [(2, 19, 7)]
+    if TO_CPU
+    else [(1,), (1024, 1024), (20, 320, 15), (16, 128, 64, 60), (16, 7, 57, 32, 29)]
```
Comment (iclementine) [inline tests/accuracy_utils.py:37]: Add a 0-d shape for pointwise operation to test that 0-d works as expected? Although we now flatten them as vectors of shape `(1, )`.

File: tests/accuracy_utils.py [hunk 3]
```diff
@@ -26,37 +26,83 @@
...
-    (16, 1024, 256),
-    (16, 7, 128, 64, 64),
-    (20, 320, 15),
-]
-DIMS = [[0], [-2], [2], [0, 2], [2, 1], [0, -1, 1]]
-TILE_DIMS = [(2,), (2, 0), (0, 2), (2, 2), (2, 2, 2), (2, 2, 2, 2)]
-REPEAT_SIZES = [(2, 3, 4, 5), (2, 0, 4, 5)]
+REDUCTION_SHAPES = [(2, 32)] if TO_CPU else [(1, 2), (4096, 256), (200, 40999, 3)]
+REDUCTION_SMALL_SHAPES = [(1, 32)] if TO_CPU else [(1, 2), (4096, 256), (200, 2560, 3)]
+NONZERO_SHAPES = [(2, 32)] if TO_CPU else REDUCTION_SHAPES + [(2637,)]
```
Comment (Bowen12992) [inline tests/accuracy_utils.py]: There are many `OP_SHAPES` we defined here but only used once for the certain op, how about move them close to the file/lines where we use them

**Review by Bowen12992 — COMMENTED** [review]
LGTM

**Review by Bowen12992 — APPROVED** [review]

**Review by Bowen12992 — DISMISSED** [review]
LGTM. Looking forward the next PR

File: tests/test_named_ops.py [hunk 4]
```diff
@@ -18,8 +18,7 @@
 
 reduction_ops_ut_map = {
     "all": (
-        "test_accuracy_all",
-        "test_accuracy_all_dim",
+        "test_accuracy_all_without_dim",
```
Comment (iclementine) [inline tests/test_named_ops.py:21]: We can use pytest mark to group test cases that test a same operator. 
https://docs.pytest.org/en/stable/how-to/mark.html

File: tests/accuracy_utils.py (hunk 1 above)
Comment (zhzhcookie) [reply tests/accuracy_utils.py:30]: > `TO_CPU` indicates to compute the reference value on CPU, so use less and smaller sizes to reduce the runtime?

Yes, we just pass the cpu tests, and will add cpu tests to ci tests soon. We will test all shapes when releasing a version.

File: tests/accuracy_utils.py (hunk 2 above)
Comment (zhzhcookie) [reply tests/accuracy_utils.py:37]: > Add a 0-d shape for pointwise operation to test that 0-d works as expected? Although we now flatten them as vectors of shape `(1, )`.

OK. Now there are some bugs when use scalar in tests code. We will fix it soon.

File: tests/accuracy_utils.py (hunk 3 above)
Comment (zhzhcookie) [reply tests/accuracy_utils.py]: > There are many `OP_SHAPES` we defined here but only used once for the certain op, how about move them close to the file/lines where we use them

Done.

File: tests/test_unary_pointwise_ops.py [hunk 5]
```diff
@@ -303,26 +303,44 @@ def test_accuracy_isfinite(shape, dtype):
...
-@pytest.mark.parametrize("dtype", ALL_FLOAT_DTYPES + ALL_INT_DTYPES)
-@pytest.mark.parametrize("dims", DIMS)
-def test_accuracy_flip(shape, dtype, dims):
+def get_max_ndim(shape, dims):
+    max_ndim = max(len(shape), len(dims))
+    for dim in dims:
+        dim = abs(dim) + 1
+        if dim > max_ndim:
+            max_ndim = dim
+    return max_ndim
```
Comment (iclementine) [inline tests/test_unary_pointwise_ops.py]: Is this used to make a tensor have at least N dimensions, such that each axis in dims is a valid axis in the tensor?

Then if axis is not negative, the ndim is required to be `ndim >= axis + 1`; and when the axis not positive, the ndim is requires to be `ndim >= abs(axis)`.



For example, a tensor is only requires to have at least 3d to get its axis `-3`.

```text
(0, 1, 2, 3)
(-4, -3, -2, -1)
```

**Review by Bowen12992 — DISMISSED** [review]

File: tests/test_unary_pointwise_ops.py (hunk 5 above)
Comment (zhzhcookie) [reply tests/test_unary_pointwise_ops.py]: > Is this used to make a tensor have at least N dimensions, such that each axis in dims is a valid axis in the tensor?
> 
> Then if axis is not negative, the ndim is required to be `ndim >= axis + 1`; and when the axis not positive, the ndim is requires to be `ndim >= abs(axis)`.
> 
> For example, a tensor is only requires to have at least 3d to get its axis `-3`.
> 
> ```
> (0, 1, 2, 3)
> (-4, -3, -2, -1)
> ```

Done.

File: tests/test_named_ops.py (hunk 4 above)
Comment (zhzhcookie) [reply tests/test_named_ops.py:21]: > We can use pytest mark to group test cases that test a same operator. https://docs.pytest.org/en/stable/how-to/mark.html

Thanks, pytest mark will be used at 2nd ed.

**Review by Bowen12992 — APPROVED** [review]

**Pull request merged**